Spring boot jwt

ponomvrenko opened the pull request.

ponomvrenko committed:
spring-boot-starter-security dependency implemented.

ponomvrenko committed:
MapperConfig has become as an interface. SecurityConfig implementation with BCryptPasswordEncoder.

ponomvrenko committed:
Created user entity + implemented with liquibase is_delete column (for soft delete).

security:
  - added CustomUserDetailsService;

ponomvrenko committed:
SecurityConfig with getPasswordEncoder() & securityFilterChain(HttpSecurity http) was configured.

ponomvrenko committed:
Added Role entity.

ponomvrenko committed:
Added Role repository. Also defined public endpoints in SecurityConfig. Added users_roles table. liquibase changes.

ponomvrenko committed:
Added @ToString.Exclude & @EqualsAndHashCode.Exclude to Set<Role> roles in User model;

ponomvrenko committed:
Modified liquibase changes for improving tables structure:
1. Names for .yaml files were changed;
2. Inserted users with liquibase file;
3. Changed users_roles table structure;
4. Changed file names into db.changelog-master.yaml;

ponomvrenko committed:
Added @EnableMethodSecurity to SecurityConfig

ponomvrenko committed:
Added jjwt dependency in pom.xml

ponomvrenko committed:
JWT implemented.

ponomvrenko committed:
Merge branch 'spring-boot-security-incorporation' into spring-boot-jwt

d1sam commented on src/main/java/com/example/onlinebookstore/controller/AuthenticationController.java:
```diff
@@ -18,8 +21,13 @@
+    private final AuthenticationService authenticationService;
+
+    @PostMapping("/login")
+    public UserLoginResponseDto login(@RequestBody UserLoginRequestDto requestDto) {
```
```suggestion
    public UserLoginResponseDto login(@RequestBody @Valid UserLoginRequestDto requestDto) {
```

d1sam commented on src/main/java/com/example/onlinebookstore/dto/user/UserLoginRequestDto.java:
```diff
@@ -0,0 +1,17 @@
+        @Email
+        String email,
+
+        @NotEmpty
```
```suggestion
        @NotBlank
```

d1sam commented on src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,51 @@
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer")) {
+            return bearerToken.substring(7);
```
try to avoid magic numbers.

d1sam commented on src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer")) {
```
try to avoid string literals, instead try to use constants.

d1sam reviewed (changes requested):
Good job, but some changes should be done!

ponomvrenko committed:
Added Valid annotation into login parameters. Added constants to JwtAuthenticationFilter. Replaced @NotEmpty annotation with @NotBlank in UserLoginRequestDto.

ponomvrenko committed:
Fixes:
- Avoid hardcoding in 8th liquibase script by using sql queries.
- Added @Repository in RoleRepository.
- Removed /api/ from public endpoints cause we already have it as servlet context path.
- Added EntityGraph that helps to avoid LazyInitializationException.

ponomvrenko committed:
Merge branch 'spring-boot-security-incorporation' into spring-boot-jwt

vovasalyha commented on src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,55 @@
+@RequiredArgsConstructor
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private static final int SUBSTRING_AFTER_BEARER = 7;
+    private static final String BEARER_TEXT = "Bearer";
```
```suggestion
    private static final String BEARER_PREFIX = "Bearer ";
```

vovasalyha commented on src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,55 @@
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader(AUTHORIZATION_HEADER);
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith(BEARER_TEXT)) {
+            return bearerToken.substring(SUBSTRING_AFTER_BEARER);
```
```suggestion
            return bearerToken.substring(BEARER_TEXT.length());
```

vovasalyha reviewed (approved):
Left few minor suggestions, but generally all good, so approving.

ponomvrenko committed:
Fixes:
- Fixed constant naming & logic.
